### PR TITLE
Include msg ids in the consensus

### PR DESCRIPTION
**Pull request by dimkouv**
Come to consensus on both sequence number and message ids.

### CODE REVIEW

File: core/services/ocr3/plugins/ccip/commit/plugin_functions.go [hunk 1]
```diff
@@ -242,17 +247,55 @@ func newMsgsConsensusForChain(
...
 	}
-	lggr.Debugw("observed message counts", "chain", chainSel, "msgSeqNums", msgSeqNums)
+	lggr.Debugw("observed message counts", "chain", chainSel, "msgSeqNumToIdCounts", msgSeqNumToIDCounts)
+
+	msgObservationsCount := make(map[model.SeqNum]int)
+	msgSeqNumToID := make(map[model.SeqNum]model.Bytes32)
+	for seqNum, idCounts := range msgSeqNumToIDCounts {
+		if len(idCounts) == 0 {
+			continue
+		}
```
Comment (makramkd) [inline core/services/ocr3/plugins/ccip/commit/plugin_functions.go]: Is this even possible given the way the code is constructed? `seqNum` is only in the map if it has a count of at least 1.

File: core/services/ocr3/plugins/ccip/commit/plugin_functions.go [hunk 2]
```diff
@@ -242,17 +247,55 @@ func newMsgsConsensusForChain(
...
 		"chain", chainSel, "fChain", fChain, "observedMsgs", len(observedMsgs))
 
-	// Reach consensus on the observed msgs sequence numbers.
-	msgSeqNums := make(map[model.SeqNum]int)
+	// First come to consensus about the (sequence number, id) pairs.
+	// For each sequence number consider correct the ID with the most votes.
+	msgSeqNumToIDCounts := make(map[model.SeqNum]map[string]int) // seqNum -> msgID -> count
 	for _, msg := range observedMsgs {
-		msgSeqNums[msg.SeqNum]++
-		// TODO: message data might be spoofed, validate the message data
```
Comment (makramkd) [inline core/services/ocr3/plugins/ccip/commit/plugin_functions.go:249]: Is this TODO still valid? Should we re-compute msg IDs? We'd need to have full msg data though, might be hard on bandwidth.

File: core/services/ocr3/plugins/ccip/commit/plugin_functions.go [hunk 3]
```diff
@@ -242,17 +247,55 @@ func newMsgsConsensusForChain(
...
+		if len(idCounts) == 0 {
+			continue
+		}
+
+		// Find the ID with the most votes for each sequence number.
+		idsSlice := make([]string, 0, len(idCounts))
+		for id := range idCounts {
+			idsSlice = append(idsSlice, id)
+		}
+		// determinism in case we have the same count for different ids
```
Comment (makramkd) [inline core/services/ocr3/plugins/ccip/commit/plugin_functions.go:274]: Same count for different id's shouldn't be possible, but if it is, would we want to log a critical error or something?

File: core/services/ocr3/plugins/ccip/internal/model/ccip.go [hunk 4]
```diff
@@ -75,6 +75,17 @@ type CCIPMsg struct {
...
 
+func (c CCIPMsg) IsValid() error {
+	if c.ID != c.Hash() {
+		return fmt.Errorf("message id does not match the computed hash")
+	}
+	return nil
+}
+
+func (c CCIPMsg) Hash() Bytes32 {
+	return c.ID // todo: hash msg fields similar to what the contract does
```
Comment (makramkd) [inline core/services/ocr3/plugins/ccip/internal/model/ccip.go:86]: Oh I see we'd compute hash offline here. Makes sense, but will we have all the data required to compute this hash in the current context?

File: core/services/ocr3/plugins/ccip/commit/plugin_functions.go (hunk 1 above)
Comment (dimkouv) [reply core/services/ocr3/plugins/ccip/commit/plugin_functions.go]: it's not, i added an error log indicating something critical is going on

File: core/services/ocr3/plugins/ccip/commit/plugin_functions.go (hunk 2 above)
Comment (dimkouv) [reply core/services/ocr3/plugins/ccip/commit/plugin_functions.go:249]: Yeah we have to re-compute msg ids on observation phase.

File: core/services/ocr3/plugins/ccip/commit/plugin_functions.go (hunk 3 above)
Comment (dimkouv) [reply core/services/ocr3/plugins/ccip/commit/plugin_functions.go:274]: Why not? e.g. we have 10 observers. 5 observed `a` and 5 observed `b`. I remember this was something also raised in the v1 audit.

File: core/services/ocr3/plugins/ccip/internal/model/ccip.go (hunk 4 above)
Comment (dimkouv) [reply core/services/ocr3/plugins/ccip/internal/model/ccip.go:86]: yeah we should, in the observation phase.
But we don't include the whole msg in observations, just the id and the seq num (`msgBaseDetails struct`)

File: core/services/ocr3/plugins/ccip/commit/plugin_functions.go (hunk 3 above)
Comment (makramkd) [reply core/services/ocr3/plugins/ccip/commit/plugin_functions.go:274]: Main reason being that msg ids are assumed to not change after chain finality. If chain finality is violated we're in a different realm (RMN curses etc.) Fine to keep though to be super defensive

File: core/services/ocr3/plugins/ccip/internal/model/ccip.go (hunk 4 above)
Comment (makramkd) [reply core/services/ocr3/plugins/ccip/internal/model/ccip.go:86]: Are we planning to include entire msgs then? Or maybe just the sub-committees that can read the msg's source chain does the offline hash compute and compares it to what was sent but other members?

**Review by makramkd — APPROVED** [review]

**Pull request merged**